### ISSUE TITLE
SyncJobStatus is not triggered for Finish state for android 13

### ISSUE THREAD

**Issue by anchita-g** [google/android-fhir#2046]
**Describe the bug**

**Component**
Core library

**To Reproduce**

**Expected behavior**

**Screenshots**

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel4a emulator]
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

**Comment by anchita-g** [comment]
@khyativyasargus  Could you please help us with the details, specifically on the steps to reproduce.

**Comment by khyativyasargus** [comment]
> @khyativyasargus Could you please help us with the details, specifically on the steps to reproduce.

Hi @anchita-g , yes sure!

1. Go to PatientListFragment into demo app source code
2. Onserve SyncJobStatus into mainActivityViewModel.pollState.collect {} for different state of sync job, put logs there
3. Run demo app in Android 13 & observe logs , you'll notice SyncJobStatus.Finished doesn't get triggered in Android 13, but it works fine for Android 12

**Comment by khyativyasargus** [comment]
Hi @anchita-g , I hope the given explanation was helpful. Let me know if there is any update on this as it's kinda high priority for Emcare.

**Comment by anchita-g** [comment]
Hi @khyativyasargus , yes we are looking into this right now. @santosh-pingle would you be able to provide some updates on this?

**Comment by santosh-pingle** [comment]
@anchita-g 
I've already updated @khyativyasargus in the weekly call that issue is reproducible and I am looking into it.

**Comment by santosh-pingle** [comment]
Actually, issue is only reproducible on android api 33 (android 13).
sync/FhirSyncWorker

```
override suspend fun doWork(): Result {
...
val job =
      CoroutineScope(Dispatchers.IO).launch {
...
if (it is SyncJobStatus.Finished || it is SyncJobStatus.Failed) {
            this@launch.cancel()
          }
...
...
```

It looks like the Finished state does not get published when the coroutine scope is cancelled. However, if you delay the coroutine scope cancellation operation by adding delay(), then the issue is not observed.

**Comment by santosh-pingle** [comment]
<img width="875" alt="Screenshot 2023-07-10 at 11 31 34 AM" src="https://github.com/google/android-fhir/assets/86107848/e27bbf37-570d-4571-ba55-a60ba34f3d7b">


In PeriodicSyncWorker, when FhirSyncWorker updates the finished state, by the time the progress observer listens to it, FhirSyncWorker's state becomes ENQUEUE.

**Comment by santosh-pingle** [comment]
@khyativyasargus  @omarismail94 @jingtang10

**Comment by santosh-pingle** [comment]
I was just checking whether the last progress update would be observed in periodic sync with simple demo changes without flow, but I didn't see the last update get observed until an extra delay was added between setProgress() and the return call from the doWork() API.
PR https://github.com/google/android-fhir/pull/2067

**Comment by santosh-pingle** [comment]
@aditya-07

**Comment by omarismail94** [comment]
Can you cherry pick the commit from: https://github.com/google/android-fhir/pull/2064/ to update the dependecies. you wll need to have Java 17 installed for this to work

**Comment by khyativyasargus** [comment]
> Can you cherry pick the commit from: #2064 to update the dependecies. you wll need to have Java 17 installed for this to work

Hi @omarismail94  We checked with this commit [#2064](https://github.com/google/android-fhir/pull/2064) with demo app, but sync Finish state isn't being observed still.

**Comment by santosh-pingle** [comment]
fixed with #2142
